### PR TITLE
Add support for rust-analyzer type & parameter hints

### PR DESCRIPTION
**Pull request by Logarithmus**
Temporary workaround for https://github.com/autozimu/LanguageClient-neovim/issues/934, until `neovim` anticonceal https://github.com/neovim/neovim/pull/9496 is finished.
ChainingHints - `Iterator<u32>`
TypeHints - `: u32`
ParameterHints - `𝑓(a, b, c)`

Demo:

image

### CODE REVIEW

**Comment by martskins** [comment]
Nice! Very clean and readable! :+1:

**Comment by Logarithmus** [comment]
@martskins ready for review I think.

File: doc/LanguageClient.txt [hunk 1]
```diff
@@ -700,6 +674,117 @@ Default: []
...
+        },
+        "rust-analyzer": {
+          "inlayHints": {
+            "enable": true,
+            "chainingHints": true
+          },
+        },
+    }
+
+3.1 rust-analyzer     *LanguageClientSettings_rust-analyzer*
```
Comment (martskins) [inline doc/LanguageClient.txt]: I think there's a misuderstanding here, `settings.json` is not for the settings of `LanguageClient-neovim` extensions, it's essentially the initialization options for the server, so what that file ends up looking wlll not be something we control. So, having said that, I don't think this is worth documenting. We have no control over which settings make sense and which don't, that's up to each specific server.

I honestly thing the whole `Settings` section isn't worth getting into. I think it's fine to give a few examples maybe, but we don't want to start documenting specifics for each server, because we don't have control over them.I would of course keep the explanation of what the settings.json file is that was already there though.

I think the confusion may arise from the fact that we are using those settings as well to enable some calls to the server and whatnot, so we are essentially using them as settings to our extensions as well, but that's purely for simplicity and to keep it transparent for the user.

**Comment by martskins** [comment]
I'm not sure the prefixes for the different hints are easy to remember. I was going back and forth between your comment here on this PR and the screenshot to understand what each hint meant. I'm thinking if maybe surrounding the type parameters with parentheses and leaving the others without a prefix would be clearer ? (and keep the pipe in between the hints)

Nice work though, and thank you for contributing!

File: doc/LanguageClient.txt (hunk 1 above)
Comment (Logarithmus) [reply doc/LanguageClient.txt]: >I think the confusion may arise from the fact that we are using those settings as well to enable some calls to the server and whatnot, so we are essentially using them as settings to our extensions as well, but that's purely for simplicity and to keep it transparent for the user.

OK, how the user is supposed to learn about the settings of our extensions? Only by looking at the source code?
If you don't want to put this info in a separate `Settings` section, then suggest where it belongs better.

Comment (martskins) [reply doc/LanguageClient.txt]: I think I'm not communicating what I want to communicate correctly. I of course don't expect the user to look at the source code to figure out about the hints. The hints are a feature offered by rust-analyzer in this case, and they are documented there, we don't have any specific toggles for them because they are very specific to rust-analyzer, and we can use the initialization options provided by the user (in `settings.json`) as a config for our side of the feature anyways.

The initialization options, and therefore settings.json file, is a little bit of an obscure corner of the protocol precisely because each server "expects" different values in there. And I wouldn't want to go on the rabbit hole of adding each and every option for each and every server, that's up for each server to document.

**Comment by Logarithmus** [comment]
> I'm not sure the prefixes for the different hints are easy to remember. I was going back and forth between your comment here on this PR and the screenshot to understand what each hint meant. I'm thinking if maybe surrounding the type parameters with parentheses and leaving the others without a prefix would be clearer ? (and keep the pipe in between the hints)

The problem is that Github interpreted those arrows as emojis.

My logic behind those prefixes is quite simple (hopefully, it makes sense):
 - left arrow symbolizes assignment to a variable (from right to left);
 - down arrow symbolizes iterator chain going from top to bottom;
 - right arrow symbolizes input (into function).

I propose the following alternative scheme:
`ChainingHints` - simply `Iterator<u32>`
`TypeHints` - `: u32`
`ParameterHints` - `𝑓(a, b, c)`

What do you think?

File: doc/LanguageClient.txt (hunk 1 above)
Comment (Logarithmus) [reply doc/LanguageClient.txt]: >And I wouldn't want to go on the rabbit hole of adding each and every option for each and every server, that's up for each server to document.

Of course we aren't going to document each language server's initialization options. The purpose of `Settings` docs section is purely for things like toggle displaying different kinds of inlay type hints, customize prefixes for them, etc. (the things which are managed on our side, not the language server's).

Another solution would be to move these toggles to vim variables of the form `g:LanguageClient_settings_*`.

Comment (martskins) [reply doc/LanguageClient.txt]: >  toggle displaying different kinds of inlay type hints,

This is controlled by the server initialization options. We merely react to the value of the setting by doing stuff client-side so that the setting that we passed on to the server makes sense.

>  customize prefixes for them

I'm honestly not sure where something like this would live in our config/docs, but I'm also not sure it's necessary (provided we get it right at some point), going a little bit on a tangent, I would honestly remove the customization of these until requested, we can of course keep it parameterized though, so that it's easier to customize the time for that eventually comes.

> Another solution would be to move these toggles to vim variables of the form g:LanguageClient_settings_*.

I mean, maybe we'll end up going to something like server specific setting on a dictionary under some sort of 
```
let g:LanguageClient_extensions = {}
let g:LanguageClient_extensions['rust-analyzer'] = { 'chainingHintsPrefix': ' - ' }
```
But I don't think there's a justification for that now, as the toggles we can control from the settings sent to the server (if the user enables them then it should be clear enough for us that they want us to display them), and the prefixes and separator I'm not convinced need customization right now.

**Comment by martskins** [comment]
> I propose the following alternative scheme:
> ChainingHints - simply Iterator<u32>
> TypeHints - : u32
> ParameterHints - 𝑓(a, b, c)
> 
> What do you think?

I think that would be better yeah. Will need to try it out myself to see if it's clearer (to me of course) but it does seem sensible at first sight.

File: doc/LanguageClient.txt (hunk 1 above)
Comment (Logarithmus) [reply doc/LanguageClient.txt]: @martskins OK, I went to https://rust-analyzer.github.io/manual.html and saw options for `rust-analyzer.inlayHint.chainingHints` and other hint kinds as well. So it's unreasonable to document them in our project.

All hint kinds in `rust-analyzer` are turned on by default. I know that you think hints are too intrusive to show them by default, but in order to provide expected and consistent behaviour we should enable them by default too. Plus `coc-rust-analyzer`, which seems to be the most popular solution for adding `rust-analyzer` support to `nvim`, does this as well.

Summing up, I'll implement this alternative scheme https://github.com/autozimu/LanguageClient-neovim/pull/1221#issuecomment-835931022, enable hints by default (if not overriden by `settings.json`), remove customization for prefixes and remove the documentation as well.

**Comment by Logarithmus** [comment]
@martskins I've performed the changes from https://github.com/autozimu/LanguageClient-neovim/pull/1221#discussion_r629232168

**Comment by martskins** [comment]
So I've given this a try, there's a couple things that I noticed. 

The first one, is that if you have nested function calls, then the parameter hints get confusing. See line 38 in the first screenshot (from vim) here. I'm not sure how I would display this, maybe something like `f(reader: f(inner))` ? Not entirely convinced, but something like that.

The other thing is that this seems to be hitting a limit in the number of virtual texts. See line 46 in the first screenshot, that same line has a hint in the second screenshot (from VSCode).

Lastly, I think it would be nice to prefix the type parameters with the name of the variable, much like coc.nvim does. So instead of showing `: Sender<Call>, Receiver<Call>`, it would be much nicer to show something like this `tx: Sender<Call>, rx: Receiver<Call>`. I suppose this is something we can get from rust-analyzer too, but I haven't checked.

On your comment about displaying them by default, I agree. If rust-analyzer enables them by default then it makes perfect sense for us to do the same.

Again thank you for working on this, much appreciated.

Screenshot_2021-05-13_09-22-00
Screenshot_2021-05-13_09-22-22

**Comment by Logarithmus** [comment]
@martskins I'm not planning to work on this PR anymore, because neovim 0.5.0 was finally released and someday (when finally have free time) I'll migrate my config to its built-in language server support & `https://github.com/hrsh7th/nvim-compe`. If anyone interested in continuing this PR, feel free to open your own one based on mine.
